src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs

```diff
@@ -227,6 +449,95 @@ public void Dispose()
+            {
+                if (firstCol.Count != secondCol.Count)
+                {
+                    return false;
+                }
+
+                if (firstCol is IList<KeyValuePair<string, object?>> firstList && secondCol is IList<KeyValuePair<string, object?>> secondList)
+                {
+                    int count = firstList.Count;
+                    for (int i = 0; i < count; i++)
+                    {
+                        KeyValuePair<string, object?> pair1 = firstList[i];
+                        KeyValuePair<string, object?> pair2 = secondList[i];
+                        if (pair1.Key != pair2.Key || !object.Equals(pair1.Value, pair2.Value))
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            using (IEnumerator<KeyValuePair<string, object?>> e1 = tags1.GetEnumerator())
+            using (IEnumerator<KeyValuePair<string, object?>> e2 = tags2.GetEnumerator())
+            {
+                while (e1.MoveNext())
+                {
+                    KeyValuePair<string, object?> pair1 = e1.Current;
+                    if (!e2.MoveNext() || pair1.Key != e2.Current.Key || !object.Equals(pair1.Value, e2.Current.Value))
+                    {
+                        return false;
+                    }
+                }
+
+                return !e2.MoveNext();
+            }
+        }
+
+        // AddInstrument will be called when publishing the instrument (i.e. calling Instrument.Publish()).
+        private Instrument GetOrCreateInstrument<T>(Type instrumentType, string name, string? unit, string? description, IEnumerable<KeyValuePair<string, object?>>? tags, Func<Instrument> instrumentCreator)
+        {
+            List<Instrument>? instrumentList;
+
+            lock (_nonObservableInstrumentsCache)
+            {
+                if (!_nonObservableInstrumentsCache.TryGetValue(name, out instrumentList))
+                {
+                    instrumentList = new List<Instrument>();
+                    _nonObservableInstrumentsCache.Add(name, instrumentList);
+                }
+            }
+
+            Debug.Assert(instrumentList != null);
+
+            lock (instrumentList)
+            {
+                foreach (Instrument instrument in instrumentList)
+                {
+                    if (instrument.GetType() == instrumentType && instrument.Unit == unit &&
+                        instrument.Description == description && CompareTags(instrument.Tags, tags))
+                    {
+                        return instrument;
+                    }
+                }
+            }
+
+            Instrument newInstrument = instrumentCreator.Invoke();
+            lock (instrumentList)
+            {
+                instrumentList.Add(newInstrument);
+            }
```

Comment:
What prevents two threads running concurrently from both getting to this point, both creating an instrument, and both adding theirs to the list?

Reply:
That is right. If two threads using the same meter and creating instruments in same time they can run into this race condition adding to the instrument list.

Reply:
And that's ok?

Reply:
Yes, it is ok. What problem do you see here?

Reply:
I would have assumed that Instrument is supposed to be a singleton such that two different threads each getting the Instrument would end up getting the same one, but here they can end up getting a different one, and we end up storing both in the list.  And Instrument doesn't implement value equality semantics, so any code that's comparing two instruments will get false even if they're the "same".

Reply:
Won't it mean, for example, that my concern in https://github.com/dotnet/runtime/pull/86567/files#r1200580001 is actually possible, that the creation of the second duplicated instrument will end up overwriting data from the first?  Also, in https://github.com/dotnet/runtime/pull/86567/files#r1199905368 you stated the callbacks were serialized... but I'm not seeing where that happens?  The call to GetOrCreateInstrument is not under a lock, it calls instrumentCreator.Invoke() not under a lock, the instrument ctor calls Publish without a lock, and although Publish takes a lock, it releases that lock before it calls InstrumentPublished?.Invoke, so it could actually be invoked concurrently, right?

Reply:
After chatting offline we decided to check if the instrument is added to the list before adding it again.